### ISSUE TITLE
NotImplemented exception when running `nuget pack foo.csproj` on Linux

### ISSUE THREAD

**Issue by ermshiperete** [NuGet/Home#6441]
## Details about Problem

NuGet Version: 4.4.1.4656
Ubuntu 16.04.3 LTS
mono 5.4.1.6

## Detailed repro steps so we can see the same problem

1. Create a project with a `nuspec` file (my sample project is at https://gist.github.com/ermshiperete/3b9145d142eb6a3cadb795ee49de2975)

2. Build the project, then run `nuget pack NugetPackBug.csproj -Verbosity detailed`

This fails with a NotImplementedException when run on Linux (see below). On Windows things work as expected.

Creating the package works on Linux if I specify the `nuspec` file instead of the `csproj` file, but then the variables don't get substituted.

### Verbose Logs

	$ nuget pack NugetPackBug.csproj -Verbosity detailed
	NuGet Version: 4.4.1.4656
	Attempting to build package from 'NugetPackBug.csproj'.
	MSBuild auto-detection: using msbuild version '14.0' from '/usr/lib/mono/xbuild/14.0/bin'. Use option -MSBuildVersion to force nuget to use a specific version of MSBuild.
	The method or operation is not implemented.
	System.NotImplementedException: The method or operation is not implemented.
	  at (wrapper dynamic-method) System.Object:CallSite.Target (System.Runtime.CompilerServices.Closure,System.Runtime.CompilerServices.CallSite,object)
	  at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid1[T0] (System.Runtime.CompilerServices.CallSite site, T0 arg0) [0x000fa] in /build/mono-5.4.1.6/external/corefx/src/System.Linq.Expressions/src/System/Dynamic/UpdateDelegates.Generated.cs:1695 
	  at NuGet.CommandLine.ProjectFactory.ResolveTargetPath () [0x00345] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.CommandLine.ProjectFactory.BuildProject () [0x0016e] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.CommandLine.ProjectFactory.CreateBuilder (System.String basePath, NuGet.Versioning.NuGetVersion version, System.String suffix, System.Boolean buildIfNeeded, NuGet.Packaging.PackageBuilder builder) [0x00015] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.Commands.PackCommandRunner.BuildFromProjectFile (System.String path) [0x00140] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.Commands.PackCommandRunner.BuildPackage (System.String path) [0x00033] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.Commands.PackCommandRunner.BuildPackage () [0x00021] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.CommandLine.PackCommand.ExecuteCommand () [0x001fe] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.CommandLine.Command.ExecuteCommandAsync () [0x00000] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.CommandLine.Command.Execute () [0x000b7] in <085005734e3944cb8cb977189c761b03>:0 
	  at NuGet.CommandLine.Program.MainCore (System.String workingDirectory, System.String[] args) [0x001b2] in <085005734e3944cb8cb977189c761b03>:0

**Comment by zhili1208** [comment]
mono nuget.exe pack won't work on Linux since mono doesn't have msbuild on Linux, please use "dotnet pack" on Linux from https://www.microsoft.com/net/download/linux, Here is the doc: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-pack?tabs=netcore2x

**Comment by ermshiperete** [comment]
By looking at the source code it seems that nuget uses `Microsoft.Build.dll`, and that assembly is available on mono. And even if it calls `msbuild` it should be trivial to detect the platform and call `xbuild` instead.

Using `dotnet pack` instead of `nuget pack` requires changing the .csproj file to the format that dotnet expects and requires some bigger changes to our workflow and infrastructure, so that isn't really an option currently.

**Comment by zhili1208** [comment]
Microsoft.Build.dll on mono is not the same one on windows, there are some API gap there, since mono is moving from xbuild to msbuild now, I think they don't have plan to fix xbuild. @mrward Does mono on Linux have msbuild now?

**Comment by mrward** [comment]
Yes, there should be an msbuild package for Linux that includes MSBuild 15.

**Comment by mishra14** [comment]
@ermshiperete please let us know if you are still blocked by this.

**Comment by ermshiperete** [comment]
@mishra14 This is actually still a problem. Trying to force it to use MSBuild 15 also doesn't work:
    
```
$ nuget pack NugetPackBug.csproj -Verbosity detailed -MSBuildVersion 15.0

NuGet Version: 4.4.1.4656
Attempting to build package from 'NugetPackBug.csproj'.
MSBuild auto-detection: using msbuild version '15.0' from '/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/'. Use option -MSBuildVersion to force nuget to use a specific version of MSBuild.
System.IO.FileNotFoundException: 
File name: '/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/Microsoft.Build.dll'
  at (wrapper managed-to-native) System.Reflection.Assembly.LoadFrom(string,bool)
  at System.Reflection.Assembly.LoadFrom (System.String assemblyFile, System.Security.Policy.Evidence securityEvidence) [0x00000] in <2e7c1c96edae44d496118948ca617c11>:0 
  at System.Reflection.Assembly.LoadFile (System.String path, System.Security.Policy.Evidence securityEvidence) [0x0002b] in <2e7c1c96edae44d496118948ca617c11>:0 
  at System.Reflection.Assembly.LoadFile (System.String path) [0x00000] in <2e7c1c96edae44d496118948ca617c11>:0 
  at NuGet.Common.MSBuildUser.LoadAssemblies (System.String msbuildDirectory) [0x00029] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.CommandLine.ProjectFactory..ctor (System.String msbuildDirectory, System.String path, System.Collections.Generic.IDictionary`2[TKey,TValue] projectProperties) [0x00016] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.CommandLine.ProjectFactory.ProjectCreator (NuGet.Commands.PackArgs packArgs, System.String path) [0x00012] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.Commands.PackCommandRunner.BuildFromProjectFile (System.String path) [0x00059] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.Commands.PackCommandRunner.BuildPackage (System.String path) [0x00033] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.Commands.PackCommandRunner.BuildPackage () [0x00021] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.CommandLine.PackCommand.ExecuteCommand () [0x001fe] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.CommandLine.Command.ExecuteCommandAsync () [0x00000] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.CommandLine.Command.Execute () [0x000b7] in <085005734e3944cb8cb977189c761b03>:0 
  at NuGet.CommandLine.Program.MainCore (System.String workingDirectory, System.String[] args) [0x001b2] in <085005734e3944cb8cb977189c761b03>:0 ```

**Comment by mishra14** [comment]
@ermshiperete  Thanks for confirming that. I have reopened the issue.

**Comment by rsaltrelli** [comment]
Hey, guys.  I ran into the issue today.
```
Ubuntu 18.04
mono-complete 5.12.0.226
NuGet 4.6.2.5055
MSBuild 15.6.0.0
```
`nuget pack` attempts to run MSBuild 14.0 by default which gives me the `NotImplementedException`.  If I use  `-msbuildversion 15.0` then I get the `FileNotFoundException`.  I'd use `dotnet pack` but that seems to ignore my nuspec file.  Any movement on this issue or advice for workarounds?

**Comment by rsaltrelli** [comment]
Hey, guys.  I ran into the issue today.
```
Ubuntu 18.04
NuGet 4.6.2.5055
MSBuild 15.6.0.0
```
`nuget pack` attempts to run MSBuild 14.0 by default which gives me the `NotImplementedException`.  If I use  `-msbuildversion 15.0` then I get the `FileNotFoundException`.  I'd use `dotnet pack` but that seems to ignore my nuspec file.  Any movement on this issue or advice for workarounds?
